Framework/NexusGeometry/src/NexusGeometrySave.cpp

```diff
@@ -0,0 +1,407 @@
+const std::string NX_INSTRUMENT = "NXinstrument";
+const std::string NX_CHAR = "NX_CHAR";
+const std::string NX_SOURCE = "NXsource";
+const std::string NX_TRANSFORMATION = "NXtransformation";
+
+// group/attribute/dataset names
+const std::string SHORT_NAME = "short_name";
+const std::string TRANSFORMATIONS = "transformations";
+const std::string LOCAL_NAME = "local_name";
+const std::string LOCATION = "location";
+const std::string ORIENTATION = "orientation";
+const std::string DEPENDS_ON = "depends_on";
+const std::string X_PIXEL_OFFSET = "x_pixel_offset";
+const std::string Y_PIXEL_OFFSET = "y_pixel_offset";
+const std::string PIXEL_SHAPE = "pixel_shape";
+const std::string SOURCE = "source";
+const std::string SAMPLE = "sample";
+const std::string DETECTOR_NUMBER = "detector_number";
+
+// metadata
+const std::string METRES = "m";
+const std::string NAME = "name";
+const std::string UNITS = "units";
+const std::string SHAPE = "shape";
+
+// NEXUS COMPLIANT ATTRIBUTE VALUES
+
+//
+
+const H5::DataSpace H5SCALAR(H5S_SCALAR);
+
+} // namespace
+
+/*
+==============================================================================================================
+
+    Helper functions
+
+
+==============================================================================================================
+*/
+
+inline H5::StrType strTypeOfSize(const std::string &str) {
+  H5::StrType stringType(1, (size_t)str.length());
+  return stringType;
+}
+
+inline void writeStrAttributeToGroupHelper(H5::Group &grp,
+                                           const std::string &attrName,
+                                           const std::string &attrVal,
+                                           H5::DataSpace dataSpace = H5SCALAR) {
+  H5::StrType dataType = strTypeOfSize(attrVal);
+  H5::Attribute attribute = grp.createAttribute(attrName, dataType, dataSpace);
+  attribute.write(dataType, attrVal);
+}
+
+inline void
+writeStrAttributeToDataSetHelper(H5::DataSet &dSet, const std::string &attrName,
+                                 const std::string &attrVal,
+                                 H5::DataSpace dataSpace = H5SCALAR) {
+  H5::StrType dataType = strTypeOfSize(attrVal);
+  auto attribute = dSet.createAttribute(attrName, dataType, dataSpace);
+  attribute.write(dataType, attrVal);
+}
+
+inline void writeStrValueToDataSetHelper(H5::DataSet &dSet,
+                                         std::string dSetValue) {
+
+  dSet.write(dSetValue, dSet.getDataType(), H5SCALAR);
+}
+
+inline void writeDetectorNumberHelper(H5::Group &grp,
+                                      const Geometry::ComponentInfo &compInfo) {
+
+  std::vector<int> detectorIndices;
+  for (hsize_t i = compInfo.root(); i > 0; --i) {
+    if (compInfo.isDetector(i - 1)) {
+      detectorIndices.push_back(i - 1);
+    }
+  }
+
+  const hsize_t dsz = detectorIndices.size();
+
+  int rank = 1;
+  hsize_t dims[1];
+  dims[0] = dsz;
+
+  int *data = (int *)malloc(dsz * sizeof(int));
```

Comment:
Can you please use `int * data = new int[dsz]`, you should not be using `malloc`

Reply:
a `std::vector` is almost certainly better and avoids the `free` below.

Reply:
switched to std vector